### ISSUE TITLE
`fn foo<T:?>() {}` compiles despite the weird trait bound

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#42004]
Both of the following compile fine <https://is.gd/k1iXTl>:
```rust
fn foo<T:?>() {}

use std::error::Error;
fn bar<E>() -> Result<(),E>
    where E : Error + ?
{
    Ok(())
}
```

I would have expected that a lonely `?` (without a trait name) in a bound was an error.

(Thanks @tomaka on IRC for the simpler repro.)

**Comment by est31** [comment]
Interestingly, this reproduces on stable, but not on beta or nightly. Here both is an error.

**Comment by kennytm** [comment]
Fixed in #41077.

https://github.com/rust-lang/rust/blob/master/src/test/parse-fail/bound-single-question-mark.rs
